qip-0009.md

```diff
@@ -0,0 +1,529 @@
+4:    end if
+5:    newInterLink ← Copy(block.InterLink)  // Copy the existing InterLink of the block
+6:    rank ← CalculateRank(block)  // Calculate the rank of the block
+7:    for i from 0 to rank - 1 do
+8:        newInterLink[i] ← block.Hash  // Update the InterLink entries up to the calculated rank with the block's hash
+9:    end for
+10:   return MerkleRoot(newInterLink)  // Return the Merkle root of the updated InterLink
+11: end function
+
+```
+
+The functions will enable Update of interlinks. Next sections describe the Prover and Verifier design using Interlinks.
+
+## Prover and Verifier Design
+
+Benefit of using non interactive proofs is that a verifier that doesn't have the state of the blockchain can confirm with a very high probability that a transaction was included in a particular Block. Since this process is non interactive there are two independent parties that participate in this process. **Prover** and **Verifier**. Prover produces the proofs, and Verifier verifies the proof. Logic of Prover proof generation and Verifier verification of the collection of proofs is shown in sections below.
+
+Paper [1] goes over how this can done in a single chain setting. In [1] the proof is separated into two parts, suffix proofs and infix proofs. Suffix proofs show the link to the highest rank of interlink close to the Block that also satisfies the $m$ (security parameter). $m$ refers to the minimum number of Interlink blocks at a particular rank needed to generate the suffix proof. If we don't have atleast $m$ blocks from a particular rank, we default to $\mu=-2$, i.e generating the proofs with just the zone blocks. In Quai the proof generation process has additional steps since Quai already has sparse interlinks in the Hierarchy of merge mined chains named Prime and Regions.  If the highest $\mu$ in the proof is greater than $-1$ we need to show the link from the Prime to the Zone Block (Stateful Blocks in Quai) on both ends.
+
+### Definitions
+In order to describe the proof generation process few definitions of functions and terminologies that interface with the Blockchain are required. These will be listed below along with the helper methods and are used subsequently in the algorithms to generate the proof and verification of proofs.
+
+* **currentHead()** : Returns the current head of the Zone chain that the prover starts from
+* **ParentHeader(order)** : Returns the parent header of the block at a given order
+* **Number(order)** : Returns the number commited to in the header at a given order
+* **GetHeader(number/Hash, order)** : Returns the Header stored for  a given Number in Canonical Chain or Hash for a given order
+* **MerkleProof(Header, txHash)**: Returns the MerkleProof required to prove the validity of  the txHash in Header. To understand this more [Proofs in Patricia merkle tries](https://ouvrard-pierre-alain.medium.com/merkle-proof-verification-for-ethereum-patricia-tree-48f29658eec)
```

Comment:
Proof sizes can be improved with Verkle trees instead.

Reply:
Not relevant for the QIP i think, valid point